base/task.jl

```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    tasks = Task[]
+    done_mask = Bool[]
+    exception = false
+    nremaining::Int = 0
+
+    for (i, t) in enumerate(waiting_tasks)
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+        if istaskdone(t)
+            push!(done_mask, true)
+            exception |= istaskfailed(t)
+        else
+            push!(done_mask, false)
+            nremaining += 1
```

Comment:
performance nit: enumerating `waiting_tasks` might take a long time (e.g. if it is a Channel). It could thus help being more efficient in the next stage to split off the check for done-ness until that finishes being enumerated:
```suggestion
    end

    exception = false
    nremaining::Int = length(tasks)
    done_mask = falses(nremaining)
    for (i, t) in enumerate(tasks)
        if istaskdone(t)
            done_mask[i] = true
            exception |= istaskfailed(t)
            nremaining -= 1
        else
            done_mask[i] = false
```